stubs/greenlet/greenlet/_greenlet.pyi

```diff
@@ -0,0 +1,71 @@
+from collections.abc import Callable
+from contextvars import Context
+from types import FrameType, TracebackType
+from typing import Any, overload
+from typing_extensions import Literal, TypeAlias
+
+_TraceEvent: TypeAlias = Literal["switch", "throw"]
+_TraceCallback: TypeAlias = Callable[[_TraceEvent, tuple[greenlet, greenlet]], object]
+
+CLOCKS_PER_SEC: int
+GREENLET_USE_CONTEXT_VARS: bool
+GREENLET_USE_GC: bool
+GREENLET_USE_STANDARD_THREADING: bool
+GREENLET_USE_TRACING: bool
+# this is a PyCapsule, it may be used to pass the gevent C-API to another C-extension
+# there isn't a runtime type for this, since it's only an opaque wrapper around void*
+# but it's probably still better than pretending it doesn't exist, so people that need
+# to pass this around, can still pass it around without having to ignore type errors...
+_C_API: object
+
+class GreenletExit(BaseException): ...
+class error(Exception): ...
+
+class greenlet:
+    @property
+    def _stack_saved(self) -> int: ...
+    @property
+    def dead(self) -> bool: ...
+    @property
+    def gr_context(self) -> Context | None: ...
+    @gr_context.setter
+    def gr_context(self, value: Context | None) -> None: ...
+    @property
+    def gr_frame(self) -> FrameType | None: ...
+    @property
+    def parent(self) -> greenlet | None: ...
+    @parent.setter
+    def parent(self, value: greenlet | None) -> None: ...
```

Comment:
Good point, I assumed the main greenlet would have no parent, but I haven't actually verified it, the source code doesn't make it immediately obvious. I mostly have experience with `gevent`, where for every greenlet the parent would be the `Hub` which is the master greenlet in charge of the event loop, I don't recall if the `Hub` itself has a parent. I will try to find out. It's possible that it raises an Exception if there is no parent.

Either way it's seemingly correct to disallow `None` in the setter, so I will rectify that.

Reply:
Mypy/pyright might complain if the getter type doesn't match the setter type — if necessary, just add a `type: ignore` and/or a `pyright: ignore`

Reply:
Actually you're correct, I had to work around this in WebOb, I will probably need to specify a dummy descriptor protocol with `__get__` and `__set__`. Ignoring the error isn't good enough, since mypy will only use the getter signature in that case, so it will still be legal to assign `None`, which is what we would like to prevent here.

Maybe it would be a good idea to do that anyways, since none of these are real properties, it was just the easiest way to annotate that they can't be deleted and/or are read only.

Reply:
Ugh. Why don't we just annotate both the setter and getter with `greenlet` for now, given we don't actually know of a situation when it might be `None`? We can always try the more-complicated custom-descriptor solution at a later date, if somebody finds a scenario where it is actually `None` and complains about it ;)

Reply:
I just checked, the master greenlet does have `None` as it's parent. Seeing as this is the only way to trace the greenlets back to their root and identify said root we probably should annotate it correctly. (The `Hub` in `gevent` does have a parent though, which is said master greenlet, which is the one that will be running by default if you don't create any greenlets).

Reply:
Okay, let's go with your custom descriptor solution then! But please add a comment outlining the situation when it's `None` 👍